File: jblux-server/src/org/jblux/server/maps/Map.java
/**
 * File: Map.java
 *
 * @author Casey Jones
 *
 * This file is part of JBlux
 * JBlux is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JBlux is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.jblux.server.maps;

import java.util.HashMap;
import org.jblux.common.Relation;
import org.jblux.common.client.NpcData;
import org.jblux.common.items.Item;
import org.jblux.util.Coordinates;

public class Map {
    private int m_id;
    private String m_name;
    private HashMap<Coordinates, Item> m_items;
    private HashMap<Coordinates, NpcData> m_npcs;

    /* Ids of adjacent maps for convenience */
    private int map_left;
    private int map_right;
    private int map_above;
    private int map_below;

    public Map(int id, String name) {
        m_id = id;
        m_name = name;
    }

    public int getID() {
        return m_id;
    }

    public String getName() {
        return m_name;
    }

    public HashMap<Coordinates, Item> getItems() {
        return m_items;
    }

    public Item getItemAt(Coordinates c) {
        return m_items.get(c);
    }

    public void setItems(HashMap<Coordinates, Item> items) {
        m_items = items;
    }

    public HashMap<Coordinates, NpcData> getNpcs() {
        return m_npcs;
    }

    public NpcData getNpcAt(Coordinates c) {
        return m_npcs.get(c);
    }

    public void setNpcs(HashMap<Coordinates, NpcData> npcs) {
        m_npcs = npcs;
    }

    public void set_adjacent_maps(short left, short right, short above,
            short below) {
        map_left = left;
        map_right = right;
        map_above = above;
        map_below = below;
    }

    public int get_adjacent_map(Relation r) {
        int id = 0;

        if(r == Relation.LEFT)
            id = map_left;
        else if(r == Relation.RIGHT)
            id = map_right;
        else if(r == Relation.TOP)
            id = map_above;
        else if(r == Relation.BOTTOM)
            id = map_below;

        return id;
    }
}

File: new-server/src/map.c
/*
 * File: map.c
 * Author: Casey Jones
 */

#include "map.h"

static struct map_t *maps;

void init_maps()
{
    maps = db_get_all_maps();
}

struct map_t* get_map_for_id(int id)
{
    struct map_t *map = maps;
    int i = 0;
    while(map != NULL)
    {
        if(map->id == id)
        {
            break;
        }
        i++;
        *map = *(map + i);
    }
}

struct map_t* get_map_for_name(char* name)
{
    struct map_t *map = maps;
    int i = 0;
    while(map != NULL)
    {
        if(strcmp(map->name, name) == 0)
        {
            break;
        }
        i++;
        *map = *(map + i);
    }

    return map;
}

struct map_t* get_adjacent_map(struct map_t *map, enum Relation rel)
{
    struct map_t *new_map = NULL;
    int id = 0;
    switch(rel)
    {
        case LEFT:
            id = map->map_left;
            break;
        case RIGHT:
            id = map->map_right;
            break;
        case ABOVE:
            id = map->map_above;
            break;
        case BELOW:
            id = map->map_below;
            break;
        default:
            break;
    }

    if(id != 0)
    {
        new_map = get_map_for_id(id);
    }

    return new_map;
}

struct coordinates_t get_map_entrance(struct map_t *map, enum Relation rel)
{
    struct coordinates_t coords;
    switch(rel)
    {
        case LEFT:
            coords = map->left_ent;
            break;
        case RIGHT:
            coords = map->right_ent;
            break;
        case ABOVE:
            coords = map->top_ent;
            break;
        case BELOW:
            coords = map->bottom_ent;
            break;
        default:
            break;
    }

    return coords;
}

